src/System.Text.JsonLab/System/Text/Json/Utf8JsonReaderStream.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.IO;
+
+namespace System.Text.JsonLab
+{
+    public ref struct Utf8JsonReaderStream
+    {
+        private Utf8JsonReader _jsonReader;
+        private Span<byte> _span;
+        private Stream _stream;
+        private byte[] _buffer;
+        private bool _isFinalBlock;
+        private long _consumed;
+
+        const int FirstSegmentSize = 1_024; //TODO: Is this necessary?
+        const int StreamSegmentSize = 4_096;
+
+        public JsonTokenType TokenType => _jsonReader.TokenType;
+        public ReadOnlySpan<byte> Value => _jsonReader.Value;
+        public long Consumed => _consumed + _jsonReader.Consumed;
+
+        public Utf8JsonReaderStream(Stream jsonStream)
+        {
+            if (!jsonStream.CanRead)
+                JsonThrowHelper.ThrowArgumentException("Stream must be readable");
+
+            _buffer = ArrayPool<byte>.Shared.Rent(FirstSegmentSize);
+            int numberOfBytes = jsonStream.Read(_buffer, 0, FirstSegmentSize);
```

Comment:
Key scenarios would likely involve asynchronous reading, right?  That'll impact both this being in a ctor and presumably the whole type being a ref struct.

Reply:
Yes. I can't do async reads atm since the type is a ref struct. I will try to resolve async reading separately.